src/mlpack/core/data/load_image.hpp

```diff
@@ -15,15 +15,19 @@
 
 #include "image_info.hpp"
 
-#ifdef HAS_STB
+#ifdef MLPACK_HAS_STB
 
 // The definition of STB_IMAGE_IMPLEMENTATION means that the implementation will
 // be included here directly.
 #define STB_IMAGE_STATIC
 #define STB_IMAGE_IMPLEMENTATION
-#include <stb_image.h>
+#ifdef MLPACK_HAS_STB_DIR
+  #include <stb/stb_image.h>
+#else
+  #include <stb_image.h>
+#endif
```

Comment:
Instead of the `MLPACK_HAS_STB_DIR` mechanism (or in addition to it), another option is to use the `__has_include` functionality available directly in C++ compilers. This may be preferable to a bespoke solution specific to mlpack, which requires more code and hence more maintenance.

`__has_include` has long been available in many C++ compilers (eg. should be in gcc 5.x, the min gcc version for mlpack).  It's now officially part of the C++17 standard.

It's fine to use `__has_include` even in C++11/C++14 mode, as it doesn't require checking the version of the available C++ standard.
* https://en.cppreference.com/w/cpp/preprocessor/include
* https://gcc.gnu.org/onlinedocs/cpp/_005f_005fhas_005finclude.html
* https://releases.llvm.org/3.1/tools/clang/docs/LanguageExtensions.html#__has_include

As a conceptual example, the `#ifdef MLPACK_HAS_STB_DIR` guard could be replaced with:
```
#if defined __has_include
  #if __has_include(<stb_image.h>)
    #include <stb_image.h>
  #elif __has_include(<stb/stb_image.h>)
    #include <stb/stb_image.h>
  #endif
#endif
```

It may also be worthwhile to replace the `<...>` brackets with `"..."` quotes within the include statements.  The latter has a slightly different effect -- it searches local directories first (and any defined on the command line via `-I`) before the official system directories (such as /usr/include).  This is in case the user has placed `stb_image.h` locally, and it's a newer version than available in the system directories.

Reply:
I originally didn't use `__has_include` because mlpack only requires C++14, but you are right that it is possible to safely use it.  I made the changes in 80092258a4.  If `__has_include` is not available, the `MLPACK_HAS_STB_DIR` mechanism is used as before.  I also updated the documentation to point out that `MLPACK_HAS_STB_DIR` is only relevant for older compilers.